feat: Introduce checkbox component

tatianainama opened the pull request:
Closes CP-2573

This PR introduces new Checkbox component using `@react-aria/checkbox`

[Figma Component](https://www.figma.com/file/vej9BUb6WbLxN9SpOK0Uqn/Prisma-Lens---Product?node-id=1728%3A1941)

<details>
<summary> Screenshot </summary>

![image](https://user-images.githubusercontent.com/8246284/188625531-c27ec7ef-c801-4664-862a-f77d8c2a15fd.png)

</details>

tatianainama commented:
Getting the following error while building Lens:

```
/src/lib.ts → ./dist/lib.js, ./dist/lib.esm.js...
[!] (plugin rpt2) Error: /components/menu/Menu.tsx(76,5): semantic error TS2345: Argument of type 'import("/node_modules/@react-stately/menu/dist/types").MenuTriggerState' is not assignable to parameter of type 'import("/node_modules/@react-aria/combobox/node_modules/@react-stately/menu/dist/types").MenuTriggerState'.
  Property 'setOpen' is missing in type 'import("/node_modules/@react-stately/menu/dist/types").MenuTriggerState' but required in type 'import("/node_modules/@react-aria/combobox/node_modules/@react-stately/menu/dist/types").MenuTriggerState'.
src/components/menu/Menu.tsx
```

I'm guessing this is related to the work done in previous PRs, isn't it?

luanvdw commented on tailwind.config.js:
```diff
@@ -72,6 +73,10 @@ module.exports = {
           "25%": { opacity: "1" },
           "100%": { transform: "translateY(0px)", opacity: "1" },
         },
+        ripple: {
+          from: { transform: "scale(0)" },
+          to: { transform: "scale(4)", opacity: "0" },
+        },
```
This was a really nice touch!

luanvdw commented on src/components/checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,137 @@
+        className={cn("flex items-center", {
+          "flex-row-reverse": props.labelPosition === "left",
+          "cursor-pointer": !props.isReadOnly && !props.isDisabled,
+        })}
```
```suggestion
"cursor-not-allowed": props.isReadOnly || props.isDisabled,
        })}
```
Can we add a `not-allowed` cursor state for `isReadOnly` && `isDisabled `

tatianainama replied:
That's a great idea!

sampolahtinen commented:
>

sampolahtinen commented:
> Getting the following error while building Lens:
> 
> ```
> /src/lib.ts → ./dist/lib.js, ./dist/lib.esm.js...
> [!] (plugin rpt2) Error: /components/menu/Menu.tsx(76,5): semantic error TS2345: Argument of type 'import("/node_modules/@react-stately/menu/dist/types").MenuTriggerState' is not assignable to parameter of type 'import("/node_modules/@react-aria/combobox/node_modules/@react-stately/menu/dist/types").MenuTriggerState'.
>   Property 'setOpen' is missing in type 'import("/node_modules/@react-stately/menu/dist/types").MenuTriggerState' but required in type 'import("/node_modules/@react-aria/combobox/node_modules/@react-stately/menu/dist/types").MenuTriggerState'.
> src/components/menu/Menu.tsx
> ```
> 
> I'm guessing this is related to the work done in previous PRs, isn't it?

This error is so mysterious. Here is some information I [shared in slack](https://prisma-company.slack.com/archives/G01KX8T24HK/p1662121569009119) 

So this error has something unknown to do with `yarn.lock`

> Getting the following error while building Lens:
> 
> ```
> /src/lib.ts → ./dist/lib.js, ./dist/lib.esm.js...
> [!] (plugin rpt2) Error: /components/menu/Menu.tsx(76,5): semantic error TS2345: Argument of type 'import("/node_modules/@react-stately/menu/dist/types").MenuTriggerState' is not assignable to parameter of type 'import("/node_modules/@react-aria/combobox/node_modules/@react-stately/menu/dist/types").MenuTriggerState'.
>   Property 'setOpen' is missing in type 'import("/node_modules/@react-stately/menu/dist/types").MenuTriggerState' but required in type 'import("/node_modules/@react-aria/combobox/node_modules/@react-stately/menu/dist/types").MenuTriggerState'.
> src/components/menu/Menu.tsx
> ```
> 
> I'm guessing this is related to the work done in previous PRs, isn't it?

Argh this error is so mysterious. I shared some information [here](https://prisma-company.slack.com/archives/G01KX8T24HK/p1662121569009119)

It is related to `yarn.lock` file in some mysterious way. Let met quickly checkout your branch and see if this happens to me

tatianainama commented:
> Argh this error is so mysterious. I shared some information [here](https://prisma-company.slack.com/archives/G01KX8T24HK/p1662121569009119)
> 
> It is related to `yarn.lock` file in some mysterious way. Let met quickly checkout your branch and see if this happens to me

I knew I read about this somewhere, thanks for the link!
So this might be because I added a new dependency, `@react-aria/checkbox` and updated the `yarn.lock` 🤔

sampolahtinen commented:
> > Argh this error is so mysterious. I shared some information [here](https://prisma-company.slack.com/archives/G01KX8T24HK/p1662121569009119)
> > It is related to `yarn.lock` file in some mysterious way. Let met quickly checkout your branch and see if this happens to me
> 
> I knew I read about this somewhere, thanks for the link! So this might be because I added a new dependency, `@react-aria/checkbox` and updated the `yarn.lock` 🤔

Yeh exactly `it could be`. This is annoyingly flaky at the moment.

So I tried building the project using your branch and for me it works:

<img width="1131" alt="image" src="https://user-images.githubusercontent.com/19226286/188665294-43b4fd7f-cdcc-47ea-8e0c-69a0f14658d7.png">

sampolahtinen commented:
> > Argh this error is so mysterious. I shared some information [here](https://prisma-company.slack.com/archives/G01KX8T24HK/p1662121569009119)
> > It is related to `yarn.lock` file in some mysterious way. Let met quickly checkout your branch and see if this happens to me
> 
> I knew I read about this somewhere, thanks for the link! So this might be because I added a new dependency, `@react-aria/checkbox` and updated the `yarn.lock` 🤔

Yeh exactly `it could be`. This is annoyingly flaky at the moment.

So I tried building the project using your branch and for me it works:

<img width="1131" alt="image" src="https://user-images.githubusercontent.com/19226286/188665294-43b4fd7f-cdcc-47ea-8e0c-69a0f14658d7.png">

tatianainama commented:
@sampolahtinen
removing `node_modules` and installing everything again fixes the issue (as always lol)

luanvdw commented:
> 

🙌  I re-deployed the preview link (without cache) and that solved the build on Netlify as well.

luanvdw commented:
> @sampolahtinen removing `node_modules` and installing everything again fixes the issue (as always lol)

🙌  I re-deployed the preview link (without cache) and that solved the build on Netlify as well.

sampolahtinen commented:
> @sampolahtinen removing `node_modules` and installing everything again fixes the issue (as always lol)

Classic. :D thats great! So I think something like this is happening:
1. The node_modules produced from unstable lens version includes different types some aria packages
2. And even if installing with different `yarn.lock` file, these packages and types do not get updated.
3. Thus wiping node_modules, installing clean with working `yarn.lock` fixes type mappings etc

sampolahtinen commented:
> > @sampolahtinen removing `node_modules` and installing everything again fixes the issue (as always lol)
> 
> 🙌 I re-deployed the preview link (without cache) and that solved the build on Netlify as well.

This is something I cant wrap my head around. Why this keeps on occurring in Netlify. Why the fresh `node_modules` still throws errors in the next run...

sampolahtinen commented on src/components/checkbox/Checkbox.tsx:
```diff
@@ -0,0 +1,138 @@
+  useEffect(() => {
+    if (rippleRef.current) {
+      rippleRef.current.classList.remove("animate-ripple")
+      rippleRef.current.offsetWidth
```
Oh interesting! Lets add a short comment and its good :)

Found this, might be interesting  to other as well https://gist.github.com/paulirish/5d52fb081b3570c81e3a

sampolahtinen commented:
Thanks for the comment! LGTM! :)
